File: 2016/day08.rb
input = STDIN.read.lines.map(&:strip)

screen = 6.times.map { [0] * 50 }

input.each do |line|
  parts = line.split

  if parts[0] == 'rect'
    x, y = parts[1].split('x').map(&:to_i)

    x.times do |i|
      y.times do |j|
        screen[j][i] = 1
      end
    end
  end

  if line['rotate row']
    y = parts[2].split('=')[1].to_i
    offset = parts[4].to_i

    screen[y] = screen[y].cycle.lazy.drop(50 - offset).take(50).to_a
  end

  if line['rotate column']
    x = parts[2].split('=')[1].to_i
    offset = parts[4].to_i

    values = screen.transpose[x].cycle.lazy.drop(6 - offset).take(6)

    values.each.with_index do |v, i|
      screen[i][x] = v
    end
  end
end

# Part 1
puts screen.flatten.reduce(:+)

# Part 2
puts screen.map(&:join).join("\n").gsub('0',' ')

File: 2016/day08.py
import sys

input = sys.stdin.read().split("\n")
screen = [[' '] * 50 for _ in range(6)]

def t(lis):
    return list(map(list, zip(*lis)))


def format_screen():
    return "\n".join([''.join(line) for line in screen])


for i in input:
    parts = i.strip().split(' ')

    if 'rect' in i:
        w, h = map(int, parts[1].split('x'))

        for x in range(w):
            for y in range(h):
                screen[y][x] = '1'

    if 'rotate row' in i:
        y = int(parts[2].split('=')[1])
        by = int(parts[-1])

        screen[y] = screen[y][(50 - by):] + screen[y][:(50 - by)]

    if 'rotate col' in i:
        x = int(parts[2].split('=')[1])
        by = int(parts[-1])

        tmp = t(screen)
        tmp[x] = tmp[x][(6 - by):] + tmp[x][:(6 - by)]
        screen = t(tmp)


print(format_screen().count('1'))
print(format_screen())
